src/serac/physics/materials/solid_functional_material.hpp

```diff
@@ -0,0 +1,246 @@
+
+    SLIC_ERROR_ROOT_IF(bulk_modulus_ < 0.0,
+                       "Bulk modulus must be positive in the linear isotropic elasticity material model.");
+  }
+
+  /**
+   * @brief Function defining the Kirchoff stress (constitutive response)
+   *
+   * @tparam T type of the Kirchoff stress (i.e. second order tensor)
+   * @param du_dX displacement gradient with respect to the reference configuration (du_dX)
+   * @return The Kirchoff stress (det(deformation gradient) * Cauchy stress) for the constitutive model
+   */
+  template <typename T>
+  SERAC_HOST_DEVICE T operator()(const T& du_dX) const
+  {
+    auto I      = Identity<dim>();
+    auto lambda = bulk_modulus_ - (2.0 / dim) * shear_modulus_;
+    auto strain = 0.5 * (du_dX + transpose(du_dX));
+    auto stress = lambda * tr(strain) * I + 2.0 * shear_modulus_ * strain;
+    return stress;
+  }
+
+  /**
+   * @brief The density (mass per volume) of the material model
+   *
+   * @tparam dim The dimension of the problem
+   * @return The density
+   */
+  SERAC_HOST_DEVICE double density(const tensor<double, dim>& /* x */) const { return density_; }
+
+private:
+  /// Density
+  double density_;
+
+  /// Bulk modulus
+  double bulk_modulus_;
+
+  /// Shear modulus
+  double shear_modulus_;
+};
+
+/**
+ * @brief Neo-Hookean material model
+ * 
+ * @tparam dim The spatial dimension of the mesh
+ */
+template <int dim>
+class NeoHookean {
+public:
+  /**
+   * @brief Construct a new Neo-Hookean object
+   * 
+   * @param density Density of the material
+   * @param shear_modulus Shear modulus of the material
+   * @param bulk_modulus Bulk modulus of the material
+   */
+  NeoHookean(double density = 1.0, double shear_modulus = 1.0, double bulk_modulus = 1.0)
+      : density_(density), bulk_modulus_(bulk_modulus), shear_modulus_(shear_modulus)
+  {
+    SLIC_ERROR_ROOT_IF(shear_modulus_ < 0.0,
+                       "Shear modulus must be positive in the linear isotropic elasticity material model.");
+
+    SLIC_ERROR_ROOT_IF(density_ < 0.0, "Density must be positive in the linear isotropic elasticity material model.");
+
+    SLIC_ERROR_ROOT_IF(bulk_modulus_ < 0.0,
+                       "Bulk modulus must be positive in the linear isotropic elasticity material model.");
+  }
+
+  /**
+   * @brief Function defining the Kirchoff stress (constitutive response)
+   *
+   * @tparam T type of the Kirchoff stress (i.e. second order tensor)
+   * @param du_dX displacement gradient with respect to the reference configuration (du_dX)
+   * @return The Kirchoff stress (det(deformation gradient) * Cauchy stress) for the constitutive model
+   */
+  template <typename T>
+  SERAC_HOST_DEVICE T operator()(const T& du_dX) const
+  {
+    auto I         = Identity<dim>();
+    auto lambda    = bulk_modulus_ - (2.0 / dim) * shear_modulus_;
+    auto B_minus_I = du_dX * transpose(du_dX) + transpose(du_dX) + du_dX;
+
+    auto J = det(du_dX + I);
```

Comment:
it might eventually be worth implementing a special function (with a better name) for `det_I_plus(du_dX) == det(I + du_dX)`, since there can be considerable loss of precision on this when `du_dX` is small

Reply:
(note, this doesn't need to be a part of this PR)

Reply:
I'll add an issue on this.